system/lib/wasmfs/syscalls.cpp

```diff
@@ -462,21 +483,20 @@ static __wasi_fd_t doOpen(path::ParsedParent parsed,
     return -ENOTDIR;
   }
 
-  // Return an error if the file exists and O_CREAT and O_EXCL are specified.
-  if (flags & O_EXCL && flags & O_CREAT) {
-    return -EEXIST;
-  }
-
   // Note that we open the file before truncating it because some backends may
   // truncate opened files more efficiently (e.g. OPFS).
   auto openFile = std::make_shared<OpenFileState>(0, flags, child);
 
   // If O_TRUNC, truncate the file if possible.
-  if ((flags & O_TRUNC) && child->is<DataFile>()) {
-    if (fileMode & WASMFS_PERM_WRITE) {
-      child->cast<DataFile>()->locked().setSize(0);
+  if (flags & O_TRUNC) {
+    if (child->is<DataFile>()) {
+      if (fileMode & WASMFS_PERM_WRITE) {
+        child->cast<DataFile>()->locked().setSize(0);
+      } else {
+        return -EACCES;
+      }
     } else {
-      return -EACCES;
+      return -EISDIR;
```

Comment:
Maybe assert the child is not a link here?

Reply:
I'll add an assert right after we finish handling the case where the child _is_ a symlink.